### ISSUE TITLE
1.5 tooltips and menu help callbacks.

### ISSUE THREAD

**Issue by ggarra13** [fltk/fltk#941]
**Is your feature request related to a problem? Please describe.**
Currently, FLTK supports static text tooltips on widgets, but sometimes it would be useful to have a callback to display information that changes based on, for example, other settings.
Menus, which support calling callbacks, currently lack a way to have a tooltip for them or a callback to briefly mention what they do on, say, a status bar.

**I am using FLTK**
 1.4

**Describe the solution you'd like**
Add a help_callback() function to Fl_Menu_Item and to Fl_Widget.

**Describe alternatives you've considered**
Nothing, as there's, afaik, no way to do this in FLTK 1.4.

**Comment by Albrecht-S** [comment]
Thanks for your request. The only thing I can say now is that we're currently on feature freeze for FLTK 1.4, hence this request will have to wait at least until 1.4.0 has been released.

**Comment by ggarra13** [comment]
I know. That's why i put 1.5 on my summary.

**Comment by erco77** [comment]
Perhaps the `onhover()` paradigm used in html would be a good name for setting a callback. I forget the name for the "unhover" equivalent.. maybe `onblur()` or some such.

**Comment by MatthiasWM** [comment]
This is a reoccurring request. An easy API would be adding a [Fl_When](https://www.fltk.org/doc-1.4/Enumerations_8H.html#adaaa8b52b78abf3ccce4787b112e6665) flags (uchar, 5 out of 8 currently used) for tooltips. Just before a tooltip may be rendered, the callback is called with the [callback_reason](https://www.fltk.org/doc-1.4/group__fl__events.html#ga5200bf611c1c4cf6b4264e6e0b05e806) set to [FL_REASON_TOOLTIP](https://www.fltk.org/doc-1.4/Enumerations_8H.html#af10b03e83b7216797dff35f766641056) or so.

I had used and FL_TOOLTIP event in my private repo for regular widgets handle() method, but that obviously does not work with menu items.

**Comment by Albrecht-S** [comment]
> I had used an FL_TOOLTIP event in my private repo...

@MatthiasWM Is this related to [STR 2041](https://www.fltk.org/str.php?L2041) which proposes to add an FL_TOOLTIP event?

I also want to mention methods Fl_Tooltip::enter() (currently undocumented) and [Fl_Tooltip::enter_area()](https://www.fltk.org/doc-1.4/classFl__Tooltip.html#a55b4d5a9a98e69eef5716ca02a16d59e) which seem to be related but I can't tell if any of these methods would be helpful for this issue. This should be investigated, I'm just mentioning it here so it isn't forgotten.

**Comment by MatthiasWM** [comment]
I implemented this as a new event FL_TOOLTIP_EVENT. Use the event to override the current tooltip with `Fl_Tooltip::override_text()` and return 1 from the event handler. b7189192e2e31ce5ca1f2eaac2a303f9b8216ded . I will solve the `Fl_Menu_` issue in the same way by adding an `FL_BEFORE_MENU` or `FL_MENU_EVENT` or similar.

Should I rename `FL_TOOLTIP_EVENT` to `FL_BEFORE_TOOLTIP`?

**Comment by MatthiasWM** [comment]
5dd1062df53c747339b875c8cb0c13df576ad4b8 . Added `FL_BEFORE_MENU` event to all widgets derived from `Fl_Menu_`. Changed `FL_TOOLTIP_EVENT` to `FL_BEFORE_TOOLTIP` to make the name better reflect the intention of the event.

**Comment by MatthiasWM** [comment]
Not closing this because the OP also wanted a tooltip on menu items. I am not a fan. I think it clutters the GUI, but I give this a week or two to hear suggestions.

**Comment by Albrecht-S** [comment]
@MatthiasWM [STR 2041](https://www.fltk.org/str.php?L2041) looks related, the idea seems to be the same or at least very similar. Do you think that STR 2041 can be closed now?

[src/Fl_Tooltip.cxx](https://github.com/fltk/fltk/blob/b7189192e2e31ce5ca1f2eaac2a303f9b8216ded/src/Fl_Tooltip.cxx#L29) defines `DEBUG` which should very likely not be there anymore. (Tested on macOS which suppresses stdout???)

Other than that, it doesn't really do what I would expect as a user (in `test/color_chooser`). Once over the colored area, the tooltip is displayed once but doesn't change when the mouse is moved. In fact the mouse must leave the area and enter again before a new tooltip is displayed. Is this intended?

There's also a new compiler warning:
```
[177/182] Building CXX object test/CMakeFiles/color_chooser.dir/color_chooser.cxx.o
/git/fltk/master/test/color_chooser.cxx:92:7: warning: ‘virtual int Image_Box::handle(int)’ can be marked override [-Wsuggest-override]
   92 |   int handle(int event) {
      |       ^~~~~~
```

Note: this comment is based on commit b7189192e2e31ce5ca1f2eaac2a303f9b8216ded (not the latest, you are too fast)!

**Comment by MatthiasWM** [comment]
> Other than that, it doesn't really do what I would expect as a user (in `test/color_chooser`). Once over the colored area, the tooltip is displayed once but doesn't change when the mouse is moved. In fact the mouse must leave the area and enter again before a new tooltip is displayed. Is this intended?

Well, I thought about that, and it is a bit of an edge case. In the sample code, one could easily put four areas within or on top of the widget and have four different tooltips that would change from area to area, so this is not needed. Also, the override text may not be related to the mouse position at all, so the old behavior is fine.

OTOH, the user may want to have a tooltip that changes when the mouse moves, maybe even follow the mouse and show, for example, the numeric value of a graph under the mouse pointer. This is whole new feature though, I'd say. 

It's a bad usage example, so I will change the sample code so that the dynamic tooltip shows something that is not position dependent. Maybe the current time?

As for the STR, yes, thanks you for that link. I closed it. I did not use the proposed code. IIRC it required that user calls Fl_Widget::tooltip(). I think that override_text() is better, because it restores the old behavior after the tooltip is gone (set and forget...).

**Comment by Albrecht-S** [comment]
@MatthiasWM Yes, I agree that a use case where you want to change the tooltip dynamically when moving the mouse can be classified as another case (as you say, moving the mouse over a curve). What I had imagined about the color_chooser example (as I said, as a "normal user") would be that I'd get the RGB value at the mouse pointer. But, yes, this seems like another use case. It's fine as it is.

**Comment by Albrecht-S** [comment]
FYI: Meanwhile there are two warnings, related to recent changes:
```
  'virtual int Image_Box::handle(int)' can be marked override [-Wsuggest-override]
 'virtual int Dynamic_Choice::handle(int)' can be marked override [-Wsuggest-override]
```

**Comment by MatthiasWM** [comment]
Ok, both items should be fixed. Default warnings are set to pedantic now on my machine.

**Comment by ggarra13** [comment]
> Not closing this because the OP also wanted a tooltip on menu items. I am not a fan. I think it clutters the GUI, but I give this a week or two to hear suggestions.

No, no.  You misunderstood.  I wanted a callback on menu items, so, in the status bar, I could get a printout with a short menu explanation.

**Comment by MatthiasWM** [comment]
> No, no. You misunderstood. I wanted a callback on menu items, so, in the status bar, I could get a printout with a short menu explanation.

Oh, when the user rolls over the menu but does not select it yet? It's wish the current menu code wasn't so complex already. I'll take a look.

**Comment by MatthiasWM** [comment]
f4978a014997656b4592c2b3b866865f76d390ea : 

When a menu item is selected by the user and a callback is set, the callback will be called with `FL_REASON_SELECTED`. The Fl_Widget pointer is set to the widget that opened the menu, a copy of the menu item's user data. If no menu item callback is set, the callback of the managing widget is called instead.

If the `FL_MENU_CHATTY` flag is set, the menu item callback may be called for additional reasons. When a menu item is highlighted, the callback is called with `FL_REASON_GOT_FOCUS`. If it is no longer highlighted, `FL_REASON_LOST_FOCUS` is sent.
